src/builder/QuarkBuilder.sol

```diff
@@ -250,38 +262,53 @@ contract QuarkBuilder {
     }
 
     // Assert that each chain has sufficient funds to cover the max cost for that chain.
-    // NOTE: This check strategy is:
-    // 1. Accrue available fund on each chain (chain balance - max cost)
-    // 2. If total available fund >= transferIntent then pass, otherwise MaxCostTooHigh()
-    // 3. Make sure the transferIntent chain always has enough fund to cover max cost
-    function assertPaymentAffordable(
-        TransferIntent memory transferIntent,
-        Accounts.ChainAccounts[] memory chainAccountsList,
-        PaymentInfo.Payment memory payment
-    ) internal pure {
-        if (payment.isToken) {
-            uint256 totalAvailableAsset = 0;
-            for (uint256 i = 0; i < payment.maxCosts.length; ++i) {
-                uint256 paymentAssetBalanceOnChain = Accounts.sumBalances(
-                    Accounts.findAssetPositions(payment.currency, payment.maxCosts[i].chainId, chainAccountsList)
-                );
-                uint256 paymentAssetNeeded = payment.maxCosts[i].amount;
-                if (paymentAssetBalanceOnChain > paymentAssetNeeded) {
-                    totalAvailableAsset += paymentAssetBalanceOnChain - paymentAssetNeeded;
-                }
+    // Check user account can cover the cost of each actions
+    function assertActionsAffordable(Actions.Action[] memory actions, Accounts.ChainAccounts[] memory chainAccountsList)
+        internal
+        pure
+    {
+        uint256 plannedBridgeAmount = 0;
```

Comment:
This needs to be 2-pass instead of 1-pass right? Without assuming ordering of the actions, you should be summing up `plannedBridgeAmount` first, and then running the validation logic afterwards.

Reply:
It's not necessary needs to be 2-pass.. I think 🤔 
Cuz the order still crucial here as user should follow the order of bridging first then transfer.  If user transfer first before bridging it should fail.  Unless we say that the order of QuarkBuilder is not guaranteed to be correct, and expect backend or client to proactively fix the order of the actions 🤔 (not likely?) In that case we will need 2-pass to account the case we get `bridge`-> `transfer` -> `bridge`  and expect BE or Client to find the third `bridge` and know to shift it to the front.

As for our current use case we must have 'bridge' actions before 'transfer' and only one 'transfer' at the end of the actions. So the `plannedBrdigeAmount` should have accrued all bridging amount before we're reaching 'transfer'

Reply:
Yeah, the current code works because bridge actions are ordered in front of transfer actions. That's what I meant by saying "Without assuming ordering of the actions". The problem IMO is the code implicitly makes this assumption on ordering, which is prone to bugs and hurts readability. I would recommend either adjusting to 2-pass or calling out this assumption using comments.

Reply:
Ok I have pushed out a newer version for this. 
So we verify two actions separately. 
And while processing bridgeAction we keep track on palnnedBridgeAmount. 

And when processing through the transactionAction, to have better future proof, we also track the accruedCost for multiple actions cases, instead of assuming it's just one action all the time.